src/core/qgsweakrelation.cpp

```diff
@@ -17,64 +17,196 @@
 #include "qgsweakrelation.h"
 #include "qgslogger.h"
+#include "qgsproviderregistry.h"
+#include "qgsprovidermetadata.h"
 
+QgsWeakRelation::QgsWeakRelation() = default;
 
-QgsWeakRelation::QgsWeakRelation( const QString &relationId, const QString &relationName, const QgsRelation::RelationStrength strength,
+QgsWeakRelation::QgsWeakRelation( const QString &relationId, const QString &relationName, const Qgis::RelationshipStrength strength,
                                   const QString &referencingLayerId, const QString &referencingLayerName, const QString &referencingLayerSource, const QString &referencingLayerProviderKey,
-                                  const QString &referencedLayerId, const QString &referencedLayerName, const QString &referencedLayerSource, const QString &referencedLayerProviderKey,
-                                  const QList<QgsRelation::FieldPair> &fieldPairs )
+                                  const QString &referencedLayerId, const QString &referencedLayerName, const QString &referencedLayerSource, const QString &referencedLayerProviderKey )
   : mReferencingLayer( referencingLayerId, referencingLayerName, referencingLayerSource, referencingLayerProviderKey )
   , mReferencedLayer( referencedLayerId, referencedLayerName, referencedLayerSource, referencedLayerProviderKey )
   , mRelationId( relationId )
   , mRelationName( relationName )
   , mStrength( strength )
-  , mFieldPairs( fieldPairs )
 {
 }
 
-QgsRelation QgsWeakRelation::resolvedRelation( const QgsProject *project, QgsVectorLayerRef::MatchType matchType ) const
+QList< QgsRelation > QgsWeakRelation::resolvedRelations( const QgsProject *project, QgsVectorLayerRef::MatchType matchType ) const
 {
-  QgsRelation relation;
-  relation.setId( mRelationId );
-  relation.setName( mRelationName );
-  relation.setStrength( mStrength );
-  QgsVectorLayerRef referencedLayerRef { mReferencedLayer };
-  QgsMapLayer *referencedLayer { referencedLayerRef.resolveWeakly( project, matchType ) };
-  if ( referencedLayer )
-  {
-    relation.setReferencedLayer( referencedLayer->id() );
-  }
-  QgsVectorLayerRef referencingLayerRef { mReferencingLayer };
-  QgsMapLayer *referencingLayer { referencingLayerRef.resolveWeakly( project, matchType ) };
-  if ( referencingLayer )
-  {
-    relation.setReferencingLayer( referencingLayer->id() );
-  }
-  for ( const auto &fp : std::as_const( mFieldPairs ) )
+  QList< QgsRelation > res;
+
+  switch ( mCardinality )
   {
-    relation.addFieldPair( fp );
+    case Qgis::RelationshipCardinality::OneToOne:
+    case Qgis::RelationshipCardinality::OneToMany:
+    case Qgis::RelationshipCardinality::ManyToOne:
+    {
+
+      QgsRelation relation;
+      relation.setId( mRelationId );
+      relation.setName( mRelationName );
+      relation.setStrength( mStrength );
+      QgsVectorLayerRef referencedLayerRef { mReferencedLayer };
+      QgsMapLayer *referencedLayer { referencedLayerRef.resolveWeakly( project, matchType ) };
```

Comment:
There's a bunch of these which if it was up to me I'd re-write as QgsMapLayer *l = ...; instead. Feel free to ignore.